### PR TITLE
New UseHttps overload with per connection callback 

### LINKED ISSUE #33452: Accessing Kestrel state in ServerOptionsSelectionCallback
In 5.0 we added the following API so users could customize TLS settings on a per connection basis:
https://github.com/dotnet/aspnetcore/blob/52eff90fbcfca39b7eb58baad597df6a99a542b0/src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs#L257

This directly exposes SslStream's ServerOptionsSelectionCallback delegate. The problem we're running into now is that customers need access to kestrel specific state inside that callback, like the ConnectionContext, transport information (IPs), enabling/disabling client cert renegotiation, etc..

There is an internal API used by the config code that wraps ServerOptionsSelectionCallback and exposes the ConnectionContext. 
https://github.com/dotnet/aspnetcore/blob/52eff90fbcfca39b7eb58baad597df6a99a542b0/src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs#L275

That API isn't very future proof, I'm already having to modify it for client certs (https://github.com/dotnet/aspnetcore/pull/33264). Modifying it to take a specific context object might be more future proof.

_updated_
```diff
+ public static ListenOptions UseHttps(this ListenOptions listenOptions, HttpsConnectionMiddlewareOptions options);

+ public class HttpsConnectionMiddlewareOptions
+ {
+   public Func<TlsCallbackContext, ValueTask<SslServerAuthenticationOptions>> OnConnection { get; set; } // Required
+   public object? OnConnectionState { get; set; }
+   public TimeSpan HandshakeTimeout { get; set; } = (our default)
+ }

+ public sealed class TlsCallbackContext
+ {
+   // ServerOptionsSelectionCallback parameters
+   public SslStream SslStream { get; }
+   public SslClientHelloInfo ClientHelloInfo { get; }
+   public object? State { get; }
+   public CancellationToken CancellationToken { get; }
+   // Kestrel specific
+   public ConnectionContext Connection { get; }
+   public bool AllowDelayedClientCertificateNegotation { get; set; }
+ }

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes #33452, contributes to #33264 (delayed client certs).

There is no new functionality here, this takes an internal API and makes it public (and more user friendly). Customers wanted this in order to A) get access to the ConnectionContext for information like the IPs, and B) to configure delayed client cert negotiation per connection.

~~@dotnet/aspnet-api-review having to pass the callback to the TlsHandshakeCallbackOptions constructor is inconvenient usage compared to assigning it as a property. This pattern was recommended because the callback is required. I recommend we instead validate the property is set inside the UseHttps method that takes the TlsHandshakeCallbackOptions.~~ I removed the constructor.

~~Open question: Should the ServerOptionsSelectionCallback still default to AllowDelayedClientCertificateNegotation = true (new in preview6)? Or should people that want delayed client certs move to the new API?~~ Removed. The new API will be required to opt into delayed client certs.

### CODE REVIEW

**Comment by Tratcher** [comment]
*Ping*

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs [hunk 1]
```diff
@@ -259,34 +256,37 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOpt
...
-            return listenOptions.UseHttps(GetTlsOptionsAsync, state, handshakeTimeout);
-
-            async ValueTask<(SslServerAuthenticationOptions, ClientCertificateMode)> GetTlsOptionsAsync(
-                ConnectionContext connection, SslStream stream, SslClientHelloInfo clientHelloInfo, object state, CancellationToken cancellationToken)
+            return listenOptions.UseHttps(new TlsHandshakeCallbackOptions(context =>
             {
-                var tlsOptions = await serverOptionsSelectionCallback(stream, clientHelloInfo, state, cancellationToken);
-                return new (tlsOptions, ClientCertificateMode.DelayCertificate);
-            }
+                context.AllowDelayedClientCertificateNegotation = true; // TODO: do we still need this enabled by default after adding the new overload?
```
Comment (BrennanConroy) [inline src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:261]: Don't we want to discourage the usage of delayed cert negotiation? So we should have it off by default IMO.

**Review by BrennanConroy — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs (hunk 1 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:261]: If we turn it off here then it won't be usable at all from this UseHttps overload. That's probably OK.

**Comment by Tratcher** [comment]
- Rebased.
- Removed the TlsHandshakeCallbackOptions constructor.
- ServerOptionsSelectionCallback no longer sets AllowDelayedClientCertificateNegotation = true.

File: src/Servers/Kestrel/Core/src/TlsHandshakeCallbackContext.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        /// The token to monitor for cancellation requests.
+        /// </summary>
+        public CancellationToken CancellationToken { get; internal set; }
+
+        // Kestrel specific
+
+        /// <summary>
+        /// Information about an individual connection.
+        /// </summary>
+        public ConnectionContext Connection { get; internal set; } = default!;
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/TlsHandshakeCallbackContext.cs:43]: 👍 This resolves part of #31303 too.

The next major thing is to not break ALPN for HTTP/2 when this new UseHttps() overload is used. Exposing the ConfigureAlpn() is better than nothing, but I would prefer if it just works without needing to know to call that.

**Review by halter73 — APPROVED** [review]

**Pull request merged**